### PR TITLE
TagTiles: created a new component Tag Tiles.

### PR DESCRIPTION
**Pull request by NileshSingh9245**
### Release

1.0.0
https://blackminystudio.atlassian.net/browse/TPA-55

### Description

- User can select
- Can Update tag

### Screenshots
Screenshot 2024-12-17 at 3 34 20 AM

### Tips

Use this code in main.dart for testing pupose.

```Dart

class TagModel {
  String tag;
  bool isSelected;

  TagModel(this.isSelected, this.tag);
  void logValue() {
    log('$tag:$isSelected');
  }
}

class HomeScreen extends StatefulWidget {
  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  final listOfTags = [
    TagModel(false, 'Home'),
    TagModel(false, 'Rent'),
    TagModel(false, 'Games'),
    TagModel(false, 'Groceries'),
    TagModel(false, 'TataDigital'),
  ];
  @override
  Widget build(BuildContext context) {
    for (final element in listOfTags) {
      element.logValue();
    }
    final theme = Theme.of(context);
    return Scaffold(
      backgroundColor: theme.colors.light,
      body: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 40.0),
        child: Center(
          child: ListView.builder(
            itemCount: listOfTags.length,
            itemBuilder: (context, index) => TagTile(
              tagName: listOfTags[index].tag,
              onSelection: (bool value) {
                setState(() {
                  listOfTags[index].isSelected = value;
                });
              },
              onSaveTag: (String value) {
                setState(() {
                  listOfTags[index].tag = value;
                });
              },
            ),
          ),
        ),
      ),
    );
  }
}

```

---

### IMPORTANT:

- (REQUIRED) Coverage Screenshots for the tests written
- (REQUIRED) UI Screenshots for the UI implementation written

### Author preparation checklist

- [x] Current branch is rebased with latest 'Staging' branch
- [x] PR is targeting only one issue
- [x] Commit messages are descriptive, consistent, and follow the project's guidelines.
- [ ] Branch name matches with at least `<jira link reference>-<short change description>`
- [x] Primary reviewer is set as assignee to the PR
- [x] No unnecessary changes on `pubspec.lock`
- [x] All added TODOs have associated Github issue references
- [x] \*.iml and, pubsepc_overrides.yaml are not checkedin
- [x] All hard coded strings are present in constants file

### Primary reviewer checklist

- [ ] All items in the author checklist is checked or not applicable
- [ ] All files including .lock and assets are reviewed
- [ ] Reasonable UI flows/cases related to the change are manually tested
- [ ] No commented-out code or unnecessary debug statements.
- [ ] No spelling or grammatical errors in code comments or documentation.
- [ ] No sensitive data (e.g., API keys, credentials) is hardcoded or exposed.
- [ ] The branch is up-to-date with the target branch (e.g., staging).
- [ ] The code adheres to established coding standards.

### Pre-merge checklist

- [ ] Two approvals received. (Changes affecting automatic PR verification, like new linter rules, formatting rules, Github workflow configurations)

---

### CODE REVIEW

File: lib/widgets/tag_tiles.dart [hunk 1]
```diff
@@ -0,0 +1,109 @@
+import 'package:flutter/material.dart';
+import '/theme/theme.dart';
+
+class TagTile extends StatefulWidget {
+  final String tagName;
+  final bool isChecked;
+
+  TagTile({required this.tagName, this.isChecked = false});
```
Comment (satyabratanayak) [inline lib/widgets/tag_tiles.dart]: use comma for formatting

File: lib/widgets/tag_tiles.dart [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+  @override
+  _TagTileState createState() => _TagTileState();
```
Comment (satyabratanayak) [inline lib/widgets/tag_tiles.dart]: Use State<>, create a "stl" and you will get it

File: lib/widgets/tag_tiles.dart [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            TextPosition(offset: _controller.text.length),
+          );
+      }
+    });
+  }
+
+  void _onTapOutside(PointerDownEvent value) {
+    setState(() {
+      _isEditing = false;
+    });
```
Comment (quiteachaos) [inline lib/widgets/tag_tiles.dart:87]: after line 61 add `_controller.clear()`

File: lib/widgets/tag_tiles.dart [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  void _toggleEditMode() {
+    setState(() {
+      _isEditing = !_isEditing;
+    });
+  }
+
+  void _onSubmitted(String value) {
+    setState(() {
+      _isEditing = false;
+      _controller.text;
```
Comment (quiteachaos) [inline lib/widgets/tag_tiles.dart]: line is redundant; i think you mean to assign `_originalText = value`

File: lib/widgets/tag_tiles.dart [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+  void _onTapOutside(PointerDownEvent value) {
+    setState(() {
+      _isEditing = false;
+    });
+  }
+
+  void _onTap() {
+    _controller
+      ..text = _controller.text
```
Comment (quiteachaos) [inline lib/widgets/tag_tiles.dart]: ..text = _originalText

File: lib/widgets/tag_tiles.dart [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  TagTile({required this.tagName, this.isChecked = false});
+
+  @override
+  _TagTileState createState() => _TagTileState();
+}
+
+class _TagTileState extends State<TagTile> {
+  late bool _isChecked;
+  late TextEditingController _controller;
+  bool _isEditing = false;
```
Comment (quiteachaos) [inline lib/widgets/tag_tiles.dart]: add `_originalText = ""`

**Review by satyabratanayak — COMMENTED** [review]
your code is just nice but, need improvement in code format

File: lib/widgets/tag_tiles.dart (hunk 3 above)
Comment (NileshSingh9245) [reply lib/widgets/tag_tiles.dart:87]: done

File: lib/widgets/tag_tiles.dart (hunk 4 above)
Comment (NileshSingh9245) [reply lib/widgets/tag_tiles.dart]: done

File: lib/widgets/tag_tiles.dart (hunk 5 above)
Comment (NileshSingh9245) [reply lib/widgets/tag_tiles.dart]: Done

File: lib/widgets/tag_tiles.dart (hunk 6 above)
Comment (NileshSingh9245) [reply lib/widgets/tag_tiles.dart]: Done, added : `String _originalText = '';`

File: lib/widgets/tag_tiles.dart (hunk 1 above)
Comment (NileshSingh9245) [reply lib/widgets/tag_tiles.dart]: Done

File: lib/widgets/tag_tiles.dart (hunk 2 above)
Comment (NileshSingh9245) [reply lib/widgets/tag_tiles.dart]: done

**Review by quiteachaos — DISMISSED** [review]

**Review by satyabratanayak — DISMISSED** [review]

**Review by krushnakantanayak — DISMISSED** [review]

File: lib/widgets/tag_tiles.dart [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+  void _onSubmitted(String value, ThemeData theme) {
+    final trimmedValue = value.trim(); // Trim extra spaces
+    if (trimmedValue.isNotEmpty && trimmedValue != widget.tagName) {
+      widget.onSaveTag(trimmedValue);
+      ScaffoldMessenger.of(context).showSnackBar(
+        SnackBar(
+          showCloseIcon: true,
+          closeIconColor: theme.colors.contrastLight,
+          content: const Text('Tag updated successfully!'),
```
Comment (quiteachaos) [inline lib/widgets/tag_tiles.dart]: use string_constants

File: lib/widgets/tag_tiles.dart [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+            child: TextField(
+              focusNode: _focusNode,
+              autofocus: true,
+              onSubmitted: (value) => _onSubmitted(value, theme),
+              maxLength: widget.maxLength,
+              controller: _tagNameController,
+              onTapOutside: (_) => _discardChanges(),
+              decoration: const InputDecoration(
+                border: InputBorder.none,
+                counterText: '',
```
Comment (quiteachaos) [inline lib/widgets/tag_tiles.dart]: create emptyString in string constant

File: lib/widgets/tag_tiles.dart [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    _focusNode.unfocus();
+  }
+
+  void _onSubmitted(String value, ThemeData theme) {
+    final trimmedValue = value.trim(); // Trim extra spaces
+    if (trimmedValue.isNotEmpty && trimmedValue != widget.tagName) {
+      widget.onSaveTag(trimmedValue);
+      ScaffoldMessenger.of(context).showSnackBar(
+        SnackBar(
+          showCloseIcon: true,
```
Comment (quiteachaos) [inline lib/widgets/tag_tiles.dart]: why is close icon necessary here

**Review by quiteachaos — CHANGES_REQUESTED** [review]

File: lib/widgets/tag_tiles.dart (hunk 9 above)
Comment (satyabratanayak) [reply lib/widgets/tag_tiles.dart]: Will remove it

**Review by krushnakantanayak — APPROVED** [review]

**Review by quiteachaos — APPROVED** [review]

**Pull request merged**